src/libsyntax/parse/token.rs

```diff
@@ -143,16 +143,20 @@ impl Lit {
 
 pub(crate) fn ident_can_begin_expr(name: ast::Name, span: Span, is_raw: bool) -> bool {
     let ident_token = Token::new(Ident(name, is_raw), span);
+    token_can_begin_expr(&ident_token)
+}
+
+pub(crate) fn token_can_begin_expr(ident_token: &Token) -> bool {
+    let name = if let TokenKind::Ident(ident, _) = ident_token.kind {
```

Comment:
This is performing redundant work in case `name.map_or` never happens, inline it? (also a `match` would subjectively be nicer here imo).